src/GraphQL/Value.hs

```diff
@@ -44,53 +63,54 @@ instance ToJSON GraphQL.Value.Value where
   toJSON (ValueString x) = toJSON x
   toJSON (ValueEnum x) = toJSON x
   toJSON (ValueList x) = toJSON x
-  toJSON (ValueMap x) = toJSON x
+  toJSON (ValueObject x) = toJSON x
   toJSON ValueNull = Aeson.Null
 
 newtype String = String Text deriving (Eq, Ord, Show)
 
 instance ToJSON String where
   toJSON (String x) = toJSON x
 
-newtype List = List [GraphQL.Value.Value] deriving (Eq, Ord, Show)
+newtype List = List [Value] deriving (Eq, Ord, Show)
 
 makeList :: (Functor f, Foldable f, ToValue a) => f a -> List
-makeList = List . toList . map toValue
+makeList = List . Protolude.toList . map toValue
 
 
 instance ToJSON List where
   toJSON (List x) = toJSON x
 
--- XXX: This is ObjectValue [ObjectField]; ObjectField Name Value upstream.
--- XXX: GraphQL spec itself sometimes says 'map' and other times 'object', but
--- jml hasn't read 100% clearly. Let's find something and stick to it, and
--- make sure that there isn't a real distinction between to the two.
-newtype Map = Map [(Name,  GraphQL.Value.Value)] deriving (Eq, Ord, Show, Monoid)
-
+-- | A literal GraphQL object.
+--
+-- Note that https://facebook.github.io/graphql/#sec-Response calls these
+-- \"Maps\", but everywhere else in the spec refers to them as objects.
+newtype Object = Object { objectFields :: [ObjectField] } deriving (Eq, Ord, Show)
 
-mapFromList :: [(Name,  GraphQL.Value.Value)] -> Map
-mapFromList = Map
+data ObjectField = ObjectField Name Value deriving (Eq, Ord, Show)
 
--- TODO this would be nicer with a prism `_ValueMap` but don't want to
--- pull in lens as dependency.
-unionMap :: [Value] -> Either Text Value
-unionMap values = map (ValueMap . fold)  (sequence (map isValueMap values))
+makeObject :: [ObjectField] -> Maybe Object
+makeObject fields
+  | fieldNames == ordNub fieldNames = pure (Object fields)
+  | otherwise = empty
   where
-    isValueMap = \case
-      (ValueMap m) -> Right m
-      _ -> Left "non-ValueMap member"
+    fieldNames = [name | ObjectField name _ <- fields]
+
+objectFromList :: [(Name, Value)] -> Maybe Object
+objectFromList = makeObject . map (uncurry ObjectField)
 
+unionObjects :: [Object] -> Maybe Object
+unionObjects = makeObject . (=<<) objectFields
```

Comment:
Changed to pointful style:

```haskell
unionObjects objects = makeObject (objects >>= objectFields)
```

That clearer? Can use `concatMap` instead of `>>=` if you'd prefer:

```haskell
unionObjects objects = makeObject (concatMap objectFields objects)
```